### DISCUSSION TITLE
Empty page response clears out pagination data

### DISCUSSION THREAD

**Discussion by m-shojaei** [TanStack/query#5401] [Q&A] [answered]
Hi! When using `useQuery`'s pagination I wonder why fetching a page that doesn't return any items clears out the whole data, even though `keepPreviousData` is set to `true`.

Here's a [sandbox](https://codesandbox.io/p/sandbox/react-query-pagination-forked-8kzrd3?file=%2Fpages%2Findex.js%3A65%2C1) example that mimics an endpoint that does not provide a `hasMore` field and even though the first pages contain data, asking for a nonexistent page sets the data to `[]`

Is that expected? If so, how would you recommend using the pagination against such endpoints?

**Comment by TkDodo** [comment, marked as answer]
`keepPreviousData` is used to show data from the previous queryKey while you are transitioning to the new queryKey. So yes, this is expected: As soon as the query for the 3rd page resolves, it will be shown. If it has no data, it has no data :)

**Reply by m-shojaei** [reply]
Thanks for the answer. Looking back at my question I think I was mixing up pagination and infinite loading concepts unnecessarily. 😅
